sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.scala

```diff
@@ -0,0 +1,442 @@
+      return false
+    }
+    rowsReturned += batchSize
+    columnarBatch.reset()
+    columnarBatch.setNumRows(batchSize)
+
+    var i = 0
+    while (i < requiredSchema.length) {
+      val field = requiredSchema(i)
+      val toColumn = columnVectors(i)
+
+      if (requestedColIds(i) < 0) {
+        toColumn.appendNulls(batchSize)
+      } else {
+        val fromColumn = batch.cols(requestedColIds(i))
+
+        if (fromColumn.isRepeating) {
+          if (fromColumn.isNull(0)) {
+            toColumn.appendNulls(batchSize)
+          } else {
+            field.dataType match {
+              case BooleanType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0) == 1
+                toColumn.appendBooleans(batchSize, data)
+
+              case ByteType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toByte
+                toColumn.appendBytes(batchSize, data)
+              case ShortType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toShort
+                toColumn.appendShorts(batchSize, data)
+              case IntegerType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toInt
+                toColumn.appendInts(batchSize, data)
+              case LongType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0)
+                toColumn.appendLongs(batchSize, data)
+
+              case DateType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toInt
+                toColumn.appendInts(batchSize, data)
+
+              case TimestampType =>
+                val data = fromColumn.asInstanceOf[TimestampColumnVector]
+                toColumn.appendLongs(batchSize, data.time(0) * 1000L + data.nanos(0) / 1000L)
+
+              case FloatType =>
+                val data = fromColumn.asInstanceOf[DoubleColumnVector].vector(0).toFloat
+                toColumn.appendFloats(batchSize, data)
+              case DoubleType =>
+                val data = fromColumn.asInstanceOf[DoubleColumnVector].vector(0)
+                toColumn.appendDoubles(batchSize, data)
+
+              case StringType =>
+                val data = fromColumn.asInstanceOf[BytesColumnVector]
+                var index = 0
+                while (index < batchSize) {
+                  toColumn.appendByteArray(data.vector(0), data.start(0), data.length(0))
+                  index += 1
+                }
+              case BinaryType =>
+                val data = fromColumn.asInstanceOf[BytesColumnVector]
+                var index = 0
+                while (index < batchSize) {
+                  toColumn.appendByteArray(data.vector(0), data.start(0), data.length(0))
+                  index += 1
+                }
+
+              case DecimalType.Fixed(precision, scale) =>
+                val d = fromColumn.asInstanceOf[DecimalColumnVector].vector(0)
```

Comment:
Similarly, the conversion from DecimalColumnVector's HiveDecimalWritable to Spark's representation should be encapsulated in a utilities method and explained.  Note that HiveDecimalWritable does provide the possibility of converting that are faster than going first to BigDecimal.  Having one version of the code would make that easier in the future..

Reply:
It's refactored first.